modules/metrics/src/main/java/org/apache/ignite/internal/metrics/DistributionMetric.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.metrics;
+
+import static java.util.Collections.unmodifiableList;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLongArray;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Distribution metric calculates counts of measurements that gets into each bounds interval.
+ * Note, that {@link #value()} will return array length of {@code bounds.length + 1}.
+ * Last element contains count of measurements bigger than most right value of bounds.
+ */
+public class DistributionMetric extends AbstractMetric implements CompositeMetric {
+    /** Count of measurement for each bound. */
+    private final AtomicLongArray measurements;
+
+    /** Bounds of measurements. */
+    private final long[] bounds;
+
+    /** List of scalar metrics. */
+    private volatile List<Metric> scalarMetrics = null;
+
+    /**
+     * The constructor.
+     *
+     * @param name Name.
+     * @param desc Description.
+     * @param bounds Bounds of the buckets.
+     */
+    public DistributionMetric(String name, @Nullable String desc, @NotNull long[] bounds) {
+        super(name, desc);
+
+        assert bounds != null && bounds.length > 0;
+        assert isSorted(bounds);
+
+        this.bounds = bounds;
+        this.measurements = new AtomicLongArray(bounds.length + 1);
+    }
+
+    /**
+     * Check whether given array is sorted.
+     *
+     * @param arr Array to check.
+     * @return {@code True} if array sorted, {@code false} otherwise.
+     */
+    private static boolean isSorted(@NotNull long[] arr) {
+        if (arr.length < 2) {
+            return true;
+        }
+
+        for (int i = 1; i < arr.length; i++) {
+            if (arr[i - 1] > arr[i]) {
```

Comment:
What if the comparable elements are equal?

Reply:
this condition will be `false` and the method will not return `false` on current check.

Reply:
I mean that equal elements must not be allowed because it will be always empty bucket (e.g. 5_5).

Reply:
I mean that such situation must be impossible. The current implementation allows buckets like this `new long[] {10, 10, 10}` but it is doesn't make sense.

Reply:
fixed, method is remaned to `isSortedAndUnique`.